From 9abdda0d9886f3eb23bbdad8b0106e694f7f29f9 Mon Sep 17 00:00:00 2001
From: Arun Ravindran <arun.ravindran@intel.com>
Date: Fri, 9 Aug 2013 13:58:52 +0300
Subject: [PORT FROM MAIN] rild: Fix issue with broken pipe due to phone process died

BZ:92794

When phone process is dead during some MTBF testing, there
is a possibility of ril writting in to socket used to accept
connections from phone process. This could result in SIGPIPE
and crahes rild daemon.

This patch uses signal() to ignore the SIGPIPE aand also adds
extra error checking while writing to the socket.

Severity: Medium

Change-Id: Id18213115f6fd165710485e9470f08853844f3f1
Orig-Change-Id: Id11688d56c2f9b995ac299dbc6b2dda9e71962d9
Category: aosp improvement
Domain: Telephony-Telephony-FWK
Origin: internal
Upstream-Candidate: no, proprietary
Signed-off-by: Arun Ravindran <arun.ravindran@intel.com>
---
 libril/ril.cpp |    3 ++-
 rild/rild.c    |    1 +
 2 files changed, 3 insertions(+), 1 deletions(-)

diff --git a/libril/ril.cpp b/libril/ril.cpp
index b0bdfd2..e3bc237 100644
--- a/libril/ril.cpp
+++ b/libril/ril.cpp
@@ -1681,7 +1681,8 @@ blockingWrite(int fd, const void *buffer, size_t len) {
         do {
             written = write (fd, toWrite + writeOffset,
                                 len - writeOffset);
-        } while (written < 0 && ((errno == EINTR) || (errno == EAGAIN)));
+        } while (written < 0 && (errno != EPIPE)
+                && ((errno == EINTR) || (errno == EAGAIN)));
 
         if (written >= 0) {
             writeOffset += written;
diff --git a/rild/rild.c b/rild/rild.c
index fe1d2ce..f6b680a 100644
--- a/rild/rild.c
+++ b/rild/rild.c
@@ -295,6 +295,7 @@ OpenLib:
 
     RIL_register(funcs);
 
+    signal(SIGPIPE, SIG_IGN);
 done:
 
     while (true) {
-- 
1.7.4.1

